Mono.Debugging/Mono.Debugging.Evaluation/AsyncOperationManager.cs

```diff
@@ -27,215 +27,172 @@
-			else {
-				methodCall.WaitForCompleted (System.Threading.Timeout.Infinite);
+			return false;
+		}
+
+		public OperationResult<TValue> Invoke<TValue> (AsyncOperationBase<TValue> mc, int timeout)
+		{
+			if (timeout <= 0)
+				throw new ArgumentOutOfRangeException("timeout", timeout, "timeout must be greater than 0");
+
+			Task<OperationResult<TValue>> task;
+			var description = mc.Description;
+			var cts = new CancellationTokenSource ();
+			var operationData = new OperationData (mc, cts);
+			lock (currentOperations) {
+				if (disposed)
+					throw new ObjectDisposedException ("Already disposed");
+				DebuggerLoggingService.LogMessage (string.Format("Starting invoke for {0}", description));
+				task = mc.InvokeAsync (cts.Token);
+				currentOperations.Add (operationData);
 			}
 
-			lock (operationsToCancel) {
-				operationsToCancel.Remove (methodCall);
-				ST.Monitor.PulseAll (operationsToCancel);
-				if (methodCall.Aborted) {
+			bool cancelledAfterTimeout = false;
+			try {
+				if (task.Wait (timeout)) {
+					DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} succeeded in {1} ms", description, timeout));
+					return task.Result;
+				}
+				DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} timed out after {1} ms. Cancelling.", description, timeout));
+				cts.Cancel ();
+				try {
+					WaitAfterCancel (mc);
+				}
+				catch (Exception e) {
+					if (IsOperationCancelledException (e)) {
+						DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} was cancelled after timeout", description));
+						cancelledAfterTimeout = true;
+					}
+					throw;
+				}
+				DebuggerLoggingService.LogMessage (string.Format ("{0} cancelling timed out", description));
+				throw new TimeOutException ();
+			}
+			catch (Exception e) {
+				if (IsOperationCancelledException (e)) {
+					if (cancelledAfterTimeout)
+						throw new TimeOutException ();
+					DebuggerLoggingService.LogMessage (string.Format ("Invoke {0} was cancelled outside before timeout", description));
 					throw new EvaluatorAbortedException ();
 				}
+				throw;
 			}
-
-			if (!string.IsNullOrEmpty (methodCall.ExceptionMessage)) {
-				throw new Exception (methodCall.ExceptionMessage);
+			finally {
+				lock (currentOperations) {
+					currentOperations.Remove (operationData);
+				}
 			}
 		}
-		
-		public void Dispose ()
+
+
+		public event EventHandler<BusyStateEventArgs> BusyStateChanged = delegate {  };
+
+		void WaitAfterCancel (IAsyncOperationBase op)
 		{
-			Disposing = true;
-			lock (operationsToCancel) {
-				foreach (AsyncOperation op in operationsToCancel) {
-					op.InternalShutdown ();
+			var desc = op.Description;
+			DebuggerLoggingService.LogMessage (string.Format ("Waiting for cancel of invoke {0}", desc));
+			try {
+				if (!op.RawTask.Wait (ShortCancelTimeout)) {
+					try {
+						BusyStateChanged (this, new BusyStateEventArgs {IsBusy = true, Description = desc});
+						op.RawTask.Wait (LongCancelTimeout);
```

Comment:
>  if we try to abort some calculation that cannot be aborted gracefully (e.g. endless loop) the debugger may NOT to fire any signal to us - finished or aborted - and in this case we can wait forever.

That's is correct, and it is the whole point of having the debugger Busy mode. In SDB, if an invocation is in progress, we can't do anything else with the debugger. So, if the invocation can't be aborted, the only thing we can do is wait forever until the debuggee is able to end the invocation, or until the user stops the debugging session. When we enter the Busy mode, a dialog is shown to the user asking what to do (keep waiting for the call to end or stop the debugging session).

If this is not a problem for CorDebug, then you can add a flag to avoid waiting forever, but we need to keep the current behavior for SDB.